src/main/java/org/jetbrains/bsp/bazel/server/api/BuildServerImpl.java

```diff
@@ -0,0 +1,113 @@
+  private final BazelBspServerRequestHelpers serverRequestHelpers;
+
+  public BuildServerImpl(
+      BuildServerService buildServerService, BazelBspServerRequestHelpers serverRequestHelpers) {
+    this.buildServerService = buildServerService;
+    this.serverRequestHelpers = serverRequestHelpers;
+  }
+
+  @Override
+  public CompletableFuture<InitializeBuildResult> buildInitialize(
+      InitializeBuildParams initializeBuildParams) {
+    return buildServerService.buildInitialize(initializeBuildParams);
+  }
+
+  @Override
+  public void onBuildInitialized() {
+    buildServerService.onBuildInitialized();
+  }
+
+  @Override
+  public CompletableFuture<Object> buildShutdown() {
+    return buildServerService.buildShutdown();
+  }
+
+  @Override
+  public void onBuildExit() {
+    buildServerService.onBuildExit();
+  }
+
+  @Override
+  public CompletableFuture<WorkspaceBuildTargetsResult> workspaceBuildTargets() {
+    return buildServerService.workspaceBuildTargets();
+  }
+
+  @Override
+  public CompletableFuture<SourcesResult> buildTargetSources(SourcesParams sourcesParams) {
+    return serverRequestHelpers.executeCommand(
+        () -> buildServerService.buildTargetSources(sourcesParams));
+  }
+
+  @Override
+  public CompletableFuture<InverseSourcesResult> buildTargetInverseSources(
+      InverseSourcesParams inverseSourcesParams) {
+    return serverRequestHelpers.executeCommand(
+        () -> buildServerService.buildTargetInverseSources(inverseSourcesParams));
+  }
+
+  @Override
+  public CompletableFuture<DependencySourcesResult> buildTargetDependencySources(
+      DependencySourcesParams dependencySourcesParams) {
+    return serverRequestHelpers.executeCommand(
+        () -> buildServerService.buildTargetDependencySources(dependencySourcesParams));
+  }
+
+  @Override
+  public CompletableFuture<ResourcesResult> buildTargetResources(ResourcesParams resourcesParams) {
+    return serverRequestHelpers.executeCommand(
+        () -> buildServerService.buildTargetResources(resourcesParams));
+  }
+
+  @Override
+  public CompletableFuture<CompileResult> buildTargetCompile(CompileParams compileParams) {
+    return serverRequestHelpers.executeCommand(
+        () -> buildServerService.buildTargetCompile(compileParams));
+  }
+
+  @Override
+  public CompletableFuture<TestResult> buildTargetTest(TestParams testParams) {
+    return serverRequestHelpers.executeCommand(
+        () -> buildServerService.buildTargetTest(testParams));
+  }
+
+  @Override
+  public CompletableFuture<RunResult> buildTargetRun(RunParams runParams) {
+    return serverRequestHelpers.executeCommand(() -> buildServerService.buildTargetRun(runParams));
+  }
+
+  @Override
+  public CompletableFuture<CleanCacheResult> buildTargetCleanCache(
+      CleanCacheParams cleanCacheParams) {
+    return serverRequestHelpers.executeCommand(
+        () -> buildServerService.buildTargetCleanCache(cleanCacheParams));
+  }
+}
```

Comment:
I believe it's a pretty standard convention to split the logic from the API implementation file where you only have simple _@Override_s, when you implement a significant and sizable interface. It's much clearer to see the flow, plus imo it's very bad to have a ton of _@Override_s mixed with a ton of huge methods that they call or even worse, like it was before, have some of those as huge methods with nested suppliers doing _@Override_s. Also it's good to have this implementation layer that simply does the redirection and wrapping with _executeCommand_ free of all of those dependencies, those having been moved to the service.

Here it especially makes sense because the implementing methods do not simply call the equivalent methods from the service but wraps most of them like this: ``serverRequestHelpers.executeCommand(() -> buildServerService.buildTargetRun(runParams))`` – so it actually combines the behaviour of two classes. Also before, the whole team were often quite confused with some methods returning CompletableFutures, some complicated Eithers, some something else – this way it's much easier to see it and work with this when you see that these implementing methods call methods from Service that return Eithers and do this wrapping, calling _executeCommand_  from _RequestHelpers_ and return _Futures_.

Reply:
Although I mentioned this file, this is mostly related to my other comment on the service itself. Shouldn't we use this opportunity to break down the services into different logical files?

Reply:
Same as [below](https://github.com/JetBrains/bazel-bsp/pull/24#discussion_r549328902). I'd leave further breaking down of the service for later.

Reply:
Maybe we should consider introducing a command pattern?